### ISSUE TITLE
InMemoryMetricRepository does not implement PrefixMetricWriter properly

### ISSUE THREAD

**Issue by jboverfelt** [spring-projects/spring-boot#7687]
Spring Boot Version: 1.3.8.RELEASE
Operating System: CentOS 6.6
Java Version: 1.7
Project: Spring Boot Actuator

Description:

**TL;DR**:  `InMemoryMetricRepository` actually only satisfies the `MetricRepository` interface - any attempt to reset by passing a group prefix will only reset the metric at the string that is passed.

[InMemoryMetricRepository](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/repository/InMemoryMetricRepository.java) implements both `MetricRepository` and `MultiMetricRepository`. 

However, these two interfaces share a method signature through extended interfaces, namely `void reset(String metricName)` through `CounterWriter` and `void reset(String group)` through `PrefixMetricWriter`. 

Both of these interfaces are ultimately satisfied by the implementation of `reset` [here](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/metrics/repository/InMemoryMetricRepository.java#L116). 

While it satisfies the compiler, I'm not sure there's a way to implement both interfaces with one class as a result of the `reset` method signature collision. Indeed, the Redis repository implementation has two separate classes, one that implements `MetricRepository` and one that implements `MultiMetricRepository`. The Redis implementation does not exhibit this bug. 

In its current form, `InMemoryMetricRepository` simply delegates to the `remove` method on the `ConcurrentNavigableMap` inside of `SimpleInMemoryRepository`. This takes the key that is passed and attempts to remove it from the map. No attempt is made to look for all keys with that prefix. 

**Workaround**: `InMemoryMetricRepository` is able to implement the `PrefixMetricReader` interface because it has a unique method signature. This can be used to get an Iterator of all the metrics with a given prefix and then the current implementation of `reset` in `InMemoryMetricRepository` can be used for each metric.

I'm happy to create a patch to fix this, but I'm not sure how you all would want to proceed so I'll stick with opening an issue for now.

**Comment by philwebb** [comment]
Thanks for the very detailed analysis. I'm not sure that the best answer is for this, it may indeed be something we just need to live with in the 1.4.x branch.

My gut instinct is that we should probably split `InMemoryMetricRepository` into two distinct beans, even if they somehow share the underlying `map`. I've tagged this for team discussion and we'll talk about it when we next meet.

**Comment by jboverfelt** [comment]
Thanks for the quick response! I figured this would have to wait until 1.5. The workaround is OK for now. Could the documentation be updated to let people know this part of the interface isn't satisfied until 1.5?

**Comment by snicoll** [comment]
@philwebb I have a proposal fix in 9490886

**Comment by snicoll** [comment]
@jboverfelt could you please give `1.5.0.BUILD-SNAPSHOT` and try and confirm it does what you expect? We now have two separate beans, a `MetricRepository` and a `MultiMetricRepository` that share the same in-memory repository. 

I am a bit nervous about registering two beans for this so a first validation on your end should help. Also, I have hopes that @dsyer can confirm this is the right call.

**Comment by jboverfelt** [comment]
Thanks for getting to this so quickly! I'll give this a go in the next few days and get back to you.

**Comment by jboverfelt** [comment]
Hi @snicoll ! Sorry this took so long - I tested 1.5.0-BUILD-SNAPSHOT and the new InMemoryMultiMetricRepository works like a charm. Thanks again!
